### PR TITLE
Remove IGNORE_MOD_TAP_INTERRUPT_PER_KEY in favour of HOLD_ON_OTHER_KEY_PRESS_PER_KEY

### PR DESCRIPTION
**Pull request by precondition**
## Description

Deprecate IGNORE_MOD_TAP_INTERRUPT_PER_KEY as a stepping stone towards making IGNORE_MOD_TAP_INTERRUPT the new default behaviour for modtaps.

<details>
<summary>Old description from the time the PR was about setting IGNORE_MOD_TAP_INTERRUPT as the new default behaviour</summary> 
TL;DR This PR removes `IGNORE_MOD_TAP_INTERRUPT` and makes it the default behaviour for mod-taps.

If we compare mod-tap behaviour with layer-tap behaviour, we get this:

- Old default mod-tap behaviour = `HOLD_ON_OTHER_KEY_PRESS` for layer-taps
- `IGNORE_MOD_TAP_INTERRUPT` for mod-taps = default layer-tap behaviour

(Where "default" refers to no particular tap hold settings enabled for this type of dual-role key and "old" refers to any QMK version without this PR)

To make this easier, let's call the current/old default MT behaviour as `HOLD_ON_INTERRUPT` since it chooses the hold action as soon as an interrupting key is pressed. @sigprof 's `HOLD_ON_OTHER_KEY_PRESS` works exactly the same way as `HOLD_ON_INTERRUPT` except that `HOLD_ON_OTHER_KEY_PRESS` has shorter key delays, i.e. it takes the tap-or-hold decision slightly faster and sends the keyboard report to the host earlier.

Ideally, the default behaviour for layer-taps should be consistent with that of mod-taps. The best candidate for "default" behaviour is an option that takes the longest to make the tap-or-hold decision.

It is a bad idea to keep a `HOLD_ON_INTERRUPT`-like behaviour as default for modtaps (like it currently is) because this is the option that takes the shortest to make the tap-or-hold decision. This means that enabling additional options like `PERMISSIVE_HOLD` is unnoticeable because `HOLD_ON_INTERRUPT` short-circuits the decision. Users have to explicitly disable this hold-on-interrupt by defining `IGNORE_MOD_TAP_INTERRUPT` in order for extra tap-hold settings to actually take effect.

Consider the following chain of event, highlighting how `HOLD_ON_INTERRUPT`/`HOLD_ON_OTHER_KEY_PRESS` short-circuits the tap-or-hold decision (mt is released before the tapping term expired):

- mt down
- other down ; HOLD_ON_OTHER_KEY_PRESS triggers
- other up      ; PERMISSIVE_HOLD triggers
- mt up          ; IGNORE_INTERRUPT triggers

Furthermore, the doc page on tap hold settings clearly states that "IGNORE_INTERRUPT" is the default behaviour for dual-role keys:
>  ## Tap-Or-Hold Decision Modes
> 
> The code which decides between the tap and hold actions of dual-role keys supports three different modes, in increasing order of preference for the hold action:
> 
> 1. The **default** mode selects the hold action only if the dual-role key is held down longer than the tapping term. In this mode **pressing other keys while the dual-role key is held down does not influence the tap-or-hold decision.**
> 
> [...]
> The default mode gives **the most delay** (if the dual-role key is held down, this mode always waits for the whole tapping term), and the other modes may give less delay when other keys are pressed, because the hold action may be selected earlier.

This PR brings the code in alignment to what the documentation says.

These changes also simplified a great deal the tap hold documentation page that most people find very hard to understand.

Finally, this does not break any existing build (`#define IGNORE_MOD_TAP_INTERRUPT[_PER_KEY]` and `get_ignore_mod_tap_interrupt()` merely become a no-op, with no compiler error) and despite the diff looking mostly red, no loss of functionality occurs.
 </details>
 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/action_tapping.c
Comment (IsaacElenbaas) [inline quantum/action_tapping.c:195]: This should be moved to the other side of the parentheses for readability

File: docs/feature_auto_shift.md
Comment (IsaacElenbaas) [inline docs/feature_auto_shift.md:284]: This needs to be updated

**Comment by IsaacElenbaas** [comment]
With this PR and `HOLD_ON_OTHER_KEYPRESS`, rolling from mod taps under `TAPPING_TERM` produces the two tap actions unlike the old behavior when not using `IGNORE_MOD_TAP_INTERRUPT`. I assume this is an unintentional change, but if not it is an issue for Retro Shift. As referencing hold times makes little sense when they also affect whether things are shifted, it would need a workaround.

File: docs/tap_hold.md
Comment (IsaacElenbaas) [inline docs/tap_hold.md:224]: Related docs (which are currently inaccurate due to the changed behavior)

**Comment by IsaacElenbaas** [comment]
Nice job on the edge cases - you caught a lot of things I wouldn't have thought of.

File: docs/tap_hold.md
Comment (precondition) [reply docs/tap_hold.md:224]: How odd. I *distinctly* remember having edited that paragraph. I might have done an oopsie during a rebase

Comment (precondition) [reply docs/tap_hold.md:224]: It seems like you're looking at an older version (although GitHub would normally flag a suggestion on outdated text :thinking: ) because the latest changes (as of 16 jan) say the following:
?> The `PERMISSIVE_HOLD` option is not noticeable if you also enable `HOLD_ON_OTHER_KEY_PRESS` because the latter option considers both the “nested tap” and “rolling press” sequences like shown above as a hold action, not the tap action. `HOLD_ON_OTHER_KEY_PRESS` makes the Tap-Or-Hold decision earlier in the chain of key events, thus taking a precedence over `PERMISSIVE_HOLD`.

File: quantum/action_tapping.c
Comment (precondition) [reply quantum/action_tapping.c:195]: Do you mean this?:
```diff
diff --git a/quantum/action_tapping.c b/quantum/action_tapping.c
index 8ab9026937..1ca4bcf5c4 100644
--- a/quantum/action_tapping.c
+++ b/quantum/action_tapping.c
@@ -182,6 +182,7 @@ bool process_tapping(keyrecord_t *keyp) {
                         (
                             // Rolled over the two keys.
                             (
+                                tapping_key.tap.interrupted == true &&
                                 (
                                     false
 #            if defined(HOLD_ON_OTHER_KEY_PRESS) || defined(HOLD_ON_OTHER_KEY_PRESS_PER_KEY)
@@ -192,7 +193,7 @@ bool process_tapping(keyrecord_t *keyp) {
 #                endif
                                     )
 #            endif
-                                ) && tapping_key.tap.interrupted == true
+                                )
                             )
                             // Makes Retro Shift ignore the default behaviour
                             // of dual-role keys (MT and LT) for nested taps

```
You'd lose out on the early exit in the evaluation of the AND (since `tapping_key.interrupted == true` is now evaluated before `false`) and it doesn't seem that much more readable to me tbh

**Comment by precondition** [comment]
> With this PR and `HOLD_ON_OTHER_KEYPRESS`, rolling from mod taps under `TAPPING_TERM` produces the two tap actions unlike the old behavior when not using `IGNORE_MOD_TAP_INTERRUPT`. I assume this is an unintentional change, but if not it is an issue for Retro Shift. As referencing hold times makes little sense when they also affect whether things are shifted, it would need a workaround.

`HOLD_ON_OTHER_KEYPRESS` doesn't exist, the correct name is `HOLD_ON_OTHER_KEY_PRESS` (notice the underscore between KEY and PRESS). If your `config.h` contains `#define HOLD_ON_OTHER_KEYPRESS` and you still observe the default "ignore interruptions" behaviour, that shouldn't be too surprising.

The expected behaviour is described in the new tables I've added to the tap-hold docs.

Here's how rolling key presses should behave in theory:

 Time | Physical key event | Default        | `PERMISSIVE_HOLD` |  `HOLD_ON_OTHER_KEY_PRESS` |  
|------|--------------------|----------------|-------------------|----------------------------|  
| 0    | `LSFT_T(KC_A)` down|                |                   |                            |  
| 110  | `KC_B` down        |                |                   | B                          |
| 130  | `LSFT_T(KC_A)` up  | ab             | ab                | B                          |  
| 140  | `KC_B` up          | ab             | ab                | B                          |

| Time | Physical key event | Default        | `PERMISSIVE_HOLD` |  `HOLD_ON_OTHER_KEY_PRESS` |  
|------|--------------------|----------------|-------------------|----------------------------|  
| 0    | `LSFT_T(KC_A)` down|                |                   |                            |  
| 110  | `KC_B` down        |                |                   | B                          |
| 200  | `LSFT_T(KC_A)` held| B              | B                 | B                          |  
| 205  | `LSFT_T(KC_A)` up  | B              | B                 | B                          |  
| 210  | `KC_B` up          | B              | B                 | B                          |

I couldn't reproduce the bug with `#define HOLD_ON_OTHER_KEY_PRESS` so I have to assume that you simply misspelled the option when testing.

If you did in fact spell it correctly in `config.h` but simply made a mistake when writing the GitHub comment then I'd interested in hearing more about your tap hold setup.

File: docs/tap_hold.md
Comment (precondition) [reply docs/tap_hold.md:224]: See https://github.com/qmk/qmk_firmware/pull/15741#issuecomment-1015906614

File: quantum/action_tapping.c
Comment (IsaacElenbaas) [reply quantum/action_tapping.c:195]: The "rolled over the two keys" comment just makes it seem like that's already been determined to me and it took me a hot sec to find where that is actually checked. The loss of the early exit would get compiled out without `HOLD_ON_OTHER_KEY_PRESS` anyway.

Comment (precondition) [reply quantum/action_tapping.c:195]: Solved in https://github.com/qmk/qmk_firmware/pull/15741/commits/c8eff5f911fa973a5be623589aa47d5fde854206

File: docs/feature_auto_shift.md
Comment (precondition) [reply docs/feature_auto_shift.md]: Solved in https://github.com/qmk/qmk_firmware/pull/15741/commits/3a58643f6802a046dc8848572e9ff717769545d1

I find the original paragraphs rather hard to understand so let me know if I misunderstood something in my update.

**Comment by drashna** [comment]
> I've [done it before](https://github.com/qmk/qmk_firmware/pull/15307) and did it again 🙁
> 
> Looks good with Retro Shift on my end.

If it makes you feel any better, I've done similar ... more than a few times.

**Comment by precondition** [comment]
I squashed and restructured my commits so that each commit only touches a specific folder. This should be easier to follow. The underlying code changes remain the same though.

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Thanks @precondition for wadding through the `action_tapping.c` pasta and implementing this feature. The changes look good to me although I have yet to try them out on a board, but you have my approval on this PR. During review I found these issues, let me know if you disagree: 

**Unit-tests:**

- [ ] Please move the ignore_mod_tap_interrupt tests into default_mod_tap (maybe rename to default tapping?) behaviour tests, as they are the new default behaviour. 
- [ ] Altough you didn't implement the `HOLD_ON_OTHER_KEY_PRESS` feature I still would like unit-tests for this behaviour to be added as part of this PR, to increase the coverage and I think it is a good oppurtunity as you are familiar with the behavour. Would that be acceptable for you?

**Backwards compatibility:**

- [ ] The now de-funct `IGNORE_MOD_TAP_INTERRUPT_PER_KEY` and `IGNORE_MOD_TAP_INTERRUPT` defines should be removed from all keebs as they are dead code prefeerably in a seperate commit to make the review easier.

- [ ] To mimic the old behavior especially for people that prefer the old default, we need to rename `get_ignore_mod_tap_interrupt` to `get_hold_on_other_key_press` on keymap level plus inversion of logic by enabling `HOLD_ON_OTHER_KEY_PRESS` for these keymaps.

I found these keymaps to still use the now defunct `get_ignore_mod_tap_interrupt` override:

```
rg get_ignore_mod_tap_interrupt

keyboards/converter/usb_usb/keymaps/chriskopher/keymap.c
178:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/handwired/dactyl_manuform/5x6_5/keymaps/cykedev/keymap.c
198:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/bastardkb/scylla/keymaps/cykedev/keymap.c
172:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/torn/keymaps/kinesish/keymap.c
140:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/crkbd/keymaps/snowe/keymap.c
182:// bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/ergodox_ez/keymaps/stamm/keymap.c
209:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/planck/keymaps/rootiest/keymap.c
1373:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t* record) {

keyboards/planck/keymaps/adamtabrams/keymap.c
258:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

keyboards/lily58/keymaps/cykedev/keymap.c
99:bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {

users/drashna/keyrecords/tapping.c
39:__attribute__((weak)) bool get_ignore_mod_tap_interrupt(uint16_t keycode, keyrecord_t *record) {
```

File: docs/config_options.md
Comment (precondition) [reply docs/config_options.md]: Fixed in https://github.com/qmk/qmk_firmware/pull/15741/commits/d8615eb061b224551f2811154971cceb9cc41b10

**Comment by precondition** [comment]
> **Backwards compatibility:**
>
> [ ] The now de-funct IGNORE_MOD_TAP_INTERRUPT_PER_KEY and IGNORE_MOD_TAP_INTERRUPT defines should be removed from all keebs as they are dead code prefeerably in a seperate commit to make the review easier.
>
> [ ] To mimic the old behavior especially for people that prefer the old default, we need to rename get_ignore_mod_tap_interrupt to get_hold_on_other_key_press on keymap level plus inversion of logic by enabling HOLD_ON_OTHER_KEY_PRESS for these keymaps.
>
> I found these keymaps to still use the now defunct get_ignore_mod_tap_interrupt override:

@KarlK90 Doesn't that count as ”changing user code en masse”, which @fauxpark advised against, in [this comment on another PR](https://github.com/qmk/qmk_firmware/pull/15286#issuecomment-980780290)?

**Comment by KarlK90** [comment]
> @KarlK90 Doesn't that count as “changing user code en masse”, which @fauxpark advised against, in [this comment on another PR](https://github.com/qmk/qmk_firmware/pull/15286#issuecomment-980780290)?

Probably, but I would argue that 9 keymaps (1 is already commented out) to change don't count as a en masse change. But the people would have to be notified and sign off on this change though.

**Comment by precondition** [comment]
@cykedev @chriskopher @snowe2010 @stamm @adamtabrams @rootiest @Cy6erBr4in @drashna 
You are tagged here because this is a pull request that makes `IGNORE_MOD_TAP_INTERRUPT` (trigger hold action only if the key is held for longer than TAPPING_TERM) the new defaut behaviour for dual-role keys. This means that defines like `IGNORE_MOD_TAP_INTERRUPT` and functions like `get_ignore_mod_tap_interrupt` are replaced with `HOLD_ON_OTHER_KEY_PRESS` and `get_hold_on_other_key_press`. 

Your personal keymap code was adapted in a backwards-compatible way in commit https://github.com/qmk/qmk_firmware/pull/15741/commits/f31a627fd621926cde651bff4fc78c6bce7be1e2. Since this commit/update affects your personal keymap files, I need a sign-off from you, agreeing to the proposed changes in order to proceed.

**Comment by precondition** [comment]
>  The now de-funct IGNORE_MOD_TAP_INTERRUPT_PER_KEY and IGNORE_MOD_TAP_INTERRUPT defines should be removed from all keebs as they are dead code prefeerably in a seperate commit to make the review easier.

@KarlK90 Do you mean only removing it from keyboard-level config.h files as I've done in commit https://github.com/qmk/qmk_firmware/pull/15741/commits/ab78408494d878d62c0c1640ad2ada55d36f1ca1 or should I also remove them from users' keymap-level config.h files?

**Comment by cykedev** [comment]
@precondition this is fine for me

**Comment by KarlK90** [comment]
> @KarlK90 Do you mean only removing it from keyboard-level config.h files as I've done in commit [ab78408](https://github.com/qmk/qmk_firmware/commit/ab78408494d878d62c0c1640ad2ada55d36f1ca1) or should I also remove them from users' keymap-level config.h files?

As they have no usage anymore or rather are the new default, please remove them completely. A ping is not necessary here because the functionality doesn't change.

**Comment by snowe2010** [comment]
does this mean I'll need to change my keymap to include `HOLD_ON_OTHER_KEY_PRESS`?

**Comment by precondition** [comment]
> does this mean I'll need to change my keymap to include `HOLD_ON_OTHER_KEY_PRESS`?

@snowe2010 You need to add the per-key version (`HOLD_ON_OTHER_KEY_PRESS_PER_KEY`) only if you ever decide to uncomment [those lines](https://github.com/qmk/qmk_firmware/blob/f31a627fd621926cde651bff4fc78c6bce7be1e2/keyboards/crkbd/keymaps/snowe/keymap.c#L201...L209). I did not add `HOLD_ON_OTHER_KEY_PRESS_PER_KEY` to your `config.h` because you commented out the per-key settings and only kept the global IGNORE_MOD_TAP_INTERRUPT setting; which is no longer necessary as that's already the global default in this PR.

**Comment by chriskopher** [comment]
@precondition I'm fine with your changes! Thanks!

**Comment by snowe2010** [comment]
@precondition sounds good to me then! go for it!

**Comment by KarlK90** [comment]
I tried out the changes on my keyboard and everything works as expected, very well done you have my approval. Only thing left is that some `IGNORE_MOD_TAP_INTERRUPT` defines silently sneaked into develop in the meantime.
```
❯ rg IGNORE_MOD_TAP_INTERRUPT
keyboards/massdrop/ctrl/keymaps/xanimos/config.h
48:// #define IGNORE_MOD_TAP_INTERRUPT    // Makes it possible to do rolling combos (zx) with keys that convert to other keys on hold, by enforcing the TAPPING_TERM for both keys. See Mod tap interrupt for details

keyboards/handwired/dactyl_manuform/3x5_3/keymaps/dlford/config.h
28:#define IGNORE_MOD_TAP_INTERRUPT // ignore hold mod if another tap occurs within tapping term

docs/zh-cn/mod_tap.md
120:在数字及字母键上使用Mod-Tap时推荐启用 `IGNORE_MOD_TAP_INTERRUPT`，以避免在快速按下下一个键时保持功能优先级。参见[忽略Mod Tap中断](zh-cn/tap_hold.md#ignore-mod-tap-interrupt)。
```

**Review by KarlK90 — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by precondition** [comment]
I should almost somehow subscribe to keymap PRs like #16696 getting merged into develop so that I can prevent more `#define IGNORE_MOD_TAP_INTERRUPT` from sneaking in under my nose.

**Comment by drashna** [comment]
Sorry didn't mean to close! 

But came to comment this: 

A point in this favor is that ZSA, we have this enabled by default, and undef it when it's not wanted, because we've found that this is a much more natural behavior for the keys, and closer to what customers expect when using the keyboard.

**Comment by precondition** [comment]
Fixed merge conflict in keyboards/ktec/ergodone/config.h

**Comment by precondition** [comment]
The lint check fail is unrelated to the PR, see comments of https://github.com/qmk/qmk_firmware/pull/16667

**Comment by KarlK90** [comment]
Has merge conflicts after #17284 would you mind converting the tests to `EXPECT_REPORT`, `EXPECT_EMPTY_REPORT` and `EXPECT_EMPTY_REPORT`?

**Comment by precondition** [comment]
> Has merge conflicts after #17284 would you mind converting the tests to `EXPECT_REPORT`, `EXPECT_EMPTY_REPORT` and `EXPECT_EMPTY_REPORT`?

@KarlK90 Done in https://github.com/qmk/qmk_firmware/pull/15741/commits/f8bc774ef8e3af1bb0cf8c49b4bfe3153eab7147 and https://github.com/qmk/qmk_firmware/pull/15741/commits/1a0154d163ddf28deb51b6c3c8f88df1dc98cdf8

**Comment by precondition** [comment]
Another rebase to remove `get_ignore_mod_tap_interrupt` from `/keyboards/adm42/rev4/keymaps/default`.

**Comment by precondition** [comment]
Would it be better if I extracted commits purely related to `HOLD_ON_OTHER_KEY_PRESS` like https://github.com/qmk/qmk_firmware/pull/15741/commits/f64861ac2008b5c284826adcb8cfd07fb9a33fd4 and https://github.com/qmk/qmk_firmware/pull/15741/commits/9b9d411997a28bc70fe28748c98ffaceb841afa5 into another PR?

**Comment by KarlK90** [comment]
> Would it be better if I extracted commits purely related to `HOLD_ON_OTHER_KEY_PRESS` like [f64861a](https://github.com/qmk/qmk_firmware/commit/f64861ac2008b5c284826adcb8cfd07fb9a33fd4) and [9b9d411](https://github.com/qmk/qmk_firmware/commit/9b9d411997a28bc70fe28748c98ffaceb841afa5) into another PR?

From my side it is fine to keep these in this PR, as it is already reviewed and approved. Sorry for the delay again.

**Comment by KarlK90** [comment]
After re-iterating your the mod-tap interrupt PR again internally we came to a decision. The main concern was how to deal with this breaking change for users that want to keep to old behavior as much as possible, so completely rolling out the new behavior by default for **this** breaking change would give no grace period / or indication for users who do want to keep the existing behavior.

Therefore I would ask you once more for changes to the PR:

1. For a grace period (TBA. but I would say one cycle) after which `IGNORE_MOD_TAP_INTERRUPT` is phased out and becomes the default, we still want want to keep it as a marker define. This would mean that the commit which removed it would have to be reverted. `IGNORE_MOD_TAP_INTERRUPT_PER_KEY` on the other hand is deprecated for good and will produce an error.

2. A compile-time switch enables `HOLD_ON_OTHER_KEY_PRESS` if no other tapping behavior is chosen, therefore we still need `IGNORE_MOD_TAP_INTERRUPT`. A message which announces the deprecation (with information) should be added to the compilation to hopefully catch as many users as possible:

```c
#if defined(IGNORE_MOD_TAP_INTERRUPT_PER_KEY)
#    error "IGNORE_MOD_TAP_INTERRUPT_PER_KEY has been removed; the code needs to be ported to use HOLD_ON_OTHER_KEY_PRESS_PER_KEY instead."
#elif !defined(IGNORE_MOD_TAP_INTERRUPT)
#    if !defined(PERMISSIVE_HOLD) && !defined(PERMISSIVE_HOLD_PER_KEY) && !defined(HOLD_ON_OTHER_KEY_PRESS) && !defined(HOLD_ON_OTHER_KEY_PRESS_PER_KEY)
#        define HOLD_ON_OTHER_KEY_PRESS
#        pragma message "IGNORE_MOD_TAP_INTERRUPT has been deprecated and will become the new standard behavior of mod and layer taps in the future. Please use HOLD_ON_OTHER_KEY_PRESS if you want to keep the behavior for both layer and mod-taps."
#    endif
#endif
```

3. Zvecr has a [PR in the pipeline](https://github.com/qmk/qmk_firmware/pull/17393) to expand linting to keymap level so we would have to add `IGNORE_MOD_TAP_INTERRUPT_PER_KEY` to the `invalid` linting rules (`info_rules.json`) and `IGNORE_MOD_TAP_INTERRUPT` to `deprecated`.

**Comment by precondition** [comment]
> For a grace period (TBA. but I would say one cycle) after which IGNORE_MOD_TAP_INTERRUPT is phased out and becomes the default, we still want want to keep it as a marker define. This would mean that the commit which removed it would have to be reverted. IGNORE_MOD_TAP_INTERRUPT_PER_KEY on the other hand is deprecated for good and will produce an error.

You're asking to remove the heart of the PR but okay

> A compile-time switch enables HOLD_ON_OTHER_KEY_PRESS if no other tapping behavior is chosen, therefore we still need IGNORE_MOD_TAP_INTERRUPT. A message which announces the deprecation (with information) should be added to the compilation to hopefully catch as many users as possible:

I added the deprecation message in https://github.com/qmk/qmk_firmware/pull/15741/commits/d7a13ed404f8ccffb949a3b94f02594ec5ed0989 but I'm of the opinion that enabling `HOLD_ON_THE_OTHER_KEY_PRESS` in the absence of other explicitly enabled tap-or-hold modes is a bad idea. `HOLD_ON_OTHER_KEY_PRESS` applies globally to *all* dual-role keys, not just mod-taps. This means that forcing that option will change how layer-taps work and users who wish to retain the old tap-preferred/ignore-interrupt mode will have *no way* to do it. 

Leaving it unconfigured will push QMK to enable `HOLD_ON_OTHER_KEY_PRESS`, enabling `PERMISSIVE_HOLD` isn't satisfactory either as it is a different tap-or-hold mode, and explicitly enabling `HOLD_ON_OTHER_KEY_PRESS` will be no different from the no-config out-of-the-box experience wherein the layer is triggered immediately on interrupts. There is no `IGNORE_LAYER_TAP_INTERRUPT` option that people can use, and adding such an option in this PR would be a step in the opposite direction...

File: quantum/process_keycode/process_auto_shift.c
Comment (precondition) [inline quantum/process_keycode/process_auto_shift.c:404]: When rewriting this, I realised that the original code totally ignores the existence or the possibility of `PERMISSIVE_HOLD`. I'm not sure if that's an intentional omission because it's irrelevant or not.

**Comment by sigprof** [comment]
> `HOLD_ON_OTHER_KEY_PRESS` applies globally to _all_ dual-role keys, not just mod-taps.

Actually now I remembered another reason why I did not like a global `HOLD_ON_OTHER_KEY_PRESS` — it also affects the behavior of keys like `OSL()` and `OSM()` (which many people don't even consider to be dual-role keys, but they actually are implemented like that, and using a global `HOLD_ON_OTHER_KEY_PRESS` would mean that those keys would revert to the hold behavior when typing quickly due to overlaps, and this may be annoying if a chain of one shot mods is needed). So a global `HOLD_ON_OTHER_KEY_PRESS` may be a really poor replacement for the old default, and we may need to have a special case option that applies only to `MT()` if we really want to preserve the original behavior while still removing `IGNORE_MOD_TAP_INTERRUPT`.

**Comment by precondition** [comment]
Solved merge conflicts.

**Comment by precondition** [comment]
Fixed implicit `get_hold_on_other_key_press_per_key` error in commit https://github.com/qmk/qmk_firmware/pull/15741/commits/09c37be3524ba8a616d11d8e73fb90b838d81bdf
```
Compiling: quantum/action.c                                                                        quantum/action.c: In function 'process_action':
quantum/action.c:484:33: error: implicit declaration of function 'get_hold_on_other_key_press' [-Werror=implicit-function-declaration]
                                 get_hold_on_other_key_press(get_event_keycode(record->event, false), record) &&
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by drashna — APPROVED** [review]
Tested, doesn't seem to break anything. (using mod taps, layer taps, tap toggle, and oneshot mods)

**Review by KarlK90 — APPROVED** [review]
Looking forward to have `IGNORE_MOD_TAP_INTERRUPT` the default behavior in the future! Thanks for pushing this and your work.

**Comment by KarlK90** [comment]
@precondition sorry one more rebase after #17007 was merged

**Review by tzarc — APPROVED** [review]
Ticking off the director review. Will merge once conflicts are sorted.

**Review by tzarc — APPROVED** [review]
Let's try that again with the correct permissions.

**Comment by precondition** [comment]
When rebasing commit e1c457456f86b1c5024ca38b58df3f6a76d161e7, I realized that I can't just rename `get_ignore_mod_tap_interrupt` by `get_hold_on_other_key_press` and swap every `true` by `false` and vice versa because the `default` case of `get_ignore_mod_tap_interrupt` only encompassed mod-taps while the `default` case of `get_hold_on_other_key_press` encompasses all tapping keys (layer taps, mod taps, one shot mods, one shot layers, toggle tap, ...). Give me a little more time to work through this.

**Pull request merged**
